core/vm/instructions.go

```diff
@@ -440,13 +442,19 @@ func opBlockhash(pc *uint64, interpreter *EVMInterpreter, scope *ScopeContext) (
 	}
 	var upper, lower uint64
 	upper = interpreter.evm.Context.BlockNumber.Uint64()
-	if upper < 2 {
+	if upper < 257 {
 		lower = 0
 	} else {
-		lower = upper - 1
+		lower = upper - 256
 	}
 	if num64 >= lower && num64 < upper {
-		num.SetBytes(interpreter.evm.Context.GetHash(num64).Bytes())
+		chainId := interpreter.evm.ChainConfig().ChainID
+		hasher := sha3.NewLegacyKeccak256()
+		hasher.Write(chainId.Bytes())
```

Comment:
It's better to pad the `chainId` to a fixed byte length. Otherwise, if we change the `chainId`, the circuit config & vk needs to be changed accordingly.

Reply:
would 16 bits work? 2^16 = 65535 which is a sufficient upper bound.

Reply:
Currently staging is using `5343532222`. I think using `uint64` would be reasonable.